### ISSUE TITLE
Surface loss is usually very small compared with Dice

### ISSUE THREAD

**Issue by prash-p** [LIVIAETS/boundary-loss#16]
Hey, thanks for sharing the code to your work! I'm working on my own simplified implementation of this, when there is only one class present. 

However, Im getting stuck on one issue: given that predictions are mostly zeros wouldn't the surface loss always be very small because it is defined as the mean of the element wise product between the prediction and GT distance map, and this element wise product would mostly consist of 0s. So until alpha is very very small (or just 0), the overall loss is mostly dominated by the dice loss and surface loss does very little to influence learning? 

I hope this is clear, sorry for the formatting, I'm writing from my phone.

**Comment by HKervadec** [comment]
Yes your comment is spot on, actually this one reason we use some scheduling on alpha (the other reason is that it simplify the hyper-parameters search).

The predictions being only zeros are a problem known for this kind of method (as mentioned [in the rebuttal](https://openreview.net/forum?id=S1gTA5VggE), [the second version of the paper](https://arxiv.org/pdf/1812.07032.pdf), and perhaps a few Github issues but I cannot find them right now), and so that is why we do not use the loss on this own on those problems.

A few things however to consider:

- What actually matters is not necessarily the loss value, but the gradients it produces (hence mean and sum are pretty much the same for reduction)
- Those problems are less present in the multi-class problem, where the other classes are not hard-negatives of the first class.

I invite you to watch the talk I gave last July, I feel a lot of what I said there is relevant to your questions:  https://www.youtube.com/watch?v=_z6gmFlD_qE

Let me know is that wasn't clear enough, I will make a more detailed reply.

**Comment by prash-p** [comment]
Thanks for replying! 

Just a follow up question: Why do you propose using a signed distance map instead of an unsigned distance map? Did you observe a significant difference in your experiments between the two formats? I would think that, for a given image, the optimal segmentations for GDL and signed distance map are different, whereas the optimal segmentations for GDL and unsigned distance maps are the same.

**Comment by HKervadec** [comment]
> Just a follow up question: Why do you propose using a signed distance map instead of an unsigned distance map? Did you observe a significant difference in your experiments between the two formats? 

If you look at the proof of Equation (4) that I posted in https://github.com/LIVIAETS/surface-loss/issues/9#issuecomment-488840311  , you will see where the negative sign comes from and how it helps to simplify the equation.

Also, if you look at the boundary loss as purely a distance based penalty, you will notice that the gradient will be negative inside the object ; pushing up the logits and probabilities for target class during the gradient descent. Opposingly, the background pixels have a positive gradient, pushing the predicted probabilities down during SGD. 

If the distance map wasn't signed, it would push the predicted probabilities down for all pixels, be it foreground or background.

**Comment by prash-p** [comment]
Thanks!
